File: 2024-09-Septmber-GFG POTD/September 3 - Minimum umber of Deletion and insertion/Solution.py
# { Driver Code Starts
class Solution:
    def minOperations(self, str1, str2):
        m = len(str1)
        n = len(str2)
        lcs_length = self.longest_common_subsequence(str1, str2, m, n)

        # Calculate the number of deletions and insertions
        deletions = m - lcs_length
        insertions = n - lcs_length

        # Total operations needed
        return deletions + insertions

    def longest_common_subsequence(self, s1, s2, m, n):
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        # Build the dp array from bottom up
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if s1[i - 1] == s2[j - 1]:
                    dp[i][j] = 1 + dp[i - 1][j - 1]
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

        return dp[m][n]  # Length of LCS


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        s1 = input()
        s2 = input()
        ob = Solution()
        print(ob.minOperations(s1, s2))
# } Driver Code Ends

File: 2024-09-Septmber-GFG POTD/September 3 - Minimum umber of Deletion and insertion/Solution.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minOperations(string str1, string str2) {
        int m = str1.length();
        int n = str2.length();
        int lcsLength = longestCommonSubsequence(str1, str2, m, n);

        // Calculate the number of deletions and insertions
        int deletions = m - lcsLength;
        int insertions = n - lcsLength;

        // Total operations needed
        return deletions + insertions;
    }

private:
    int longestCommonSubsequence(string s1, string s2, int m, int n) {
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

        // Build the dp array from bottom up
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s1[i - 1] == s2[j - 1]) {
                    dp[i][j] = 1 + dp[i - 1][j - 1];
                } else {
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }

        return dp[m][n]; // Length of LCS
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        string s1, s2;
        cin >> s1 >> s2;
        Solution ob;
        cout << ob.minOperations(s1, s2) << endl;
    }
    return 0;
}
// } Driver Code Ends
